File: quote/Projects/PluginModel/common/SaveableProperties.vb
Imports System.ComponentModel
Imports System.Reflection

Namespace Common
    
    Public Class SavableProperties
        Implements INotifyPropertyChanged, ICloneable

        Public Event PropertyChanged As PropertyChangedEventHandler _
        Implements INotifyPropertyChanged.PropertyChanged

        Private _IsDirty As Boolean

        Public Event SavableChange As EventHandler

        <Browsable(False)> _
        Public Property Subject As Object

        <Browsable(False)>
        Public ReadOnly Property Dirty As Boolean
            Get
                Return _IsDirty
            End Get
        End Property

        Public Overridable Sub MakeDirty()
            Me._IsDirty = True
            RaiseEvent SavableChange(Me, New EventArgs())
        End Sub

        Public Overridable Sub ClearDirty()
            Me._IsDirty = False
            RaiseEvent SavableChange(Me, New EventArgs())
        End Sub

        Protected Sub AddDependent(ByVal subject As SavableProperties)
            AddHandler subject.SavableChange, AddressOf OnSavableChanged
        End Sub

        Protected Sub RemoveDependent(ByVal subject As SavableProperties)
            Dim address As EventHandler = AddressOf OnSavableChanged
            RemoveHandler subject.SavableChange, address
        End Sub

        <System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters", MessageId := "args")> _
        Protected Sub OnSavableChanged(ByVal subject As SavableProperties, _
                                       args As EventArgs) _
                                       Handles Me.SavableChange
            If subject.Dirty <> Me.Dirty Then
                Me._IsDirty = True
                RaiseEvent SavableChange(Me, New EventArgs())
            End If
        End Sub

        Public Sub SendEvents()
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs("sp"))
            MakeDirty()
        End Sub

        Public Function Clone() As Object Implements System.ICloneable.Clone
            Return Me.MemberwiseClone
        End Function

    End Class

End Namespace

File: quote/Projects/PluginModel/common/SaveableProperties.cs
namespace Model.Common
{
    using System;
    using System.Collections.Generic;
    using System.ComponentModel;
    using System.Linq;

    public abstract class SavableProperties : INotifyPropertyChanged
    {
        private bool dirty;
        
        private List<SavableProperties> dependents = new List<SavableProperties>();
        
        protected SavableProperties()
        {
        }
        
        public virtual event PropertyChangedEventHandler PropertyChanged;

        public virtual event EventHandler Dirty;

        public virtual event EventHandler Clean;
        
        [Browsable(false)]
        public bool IsDirty 
        {
            get 
            { 
                return this.dirty; 
            }
            
            set
            {
                if (value == true)
                {
                    this.dirty = true;
                    
                    if (this.Dirty != null) 
                    {
                        this.Dirty(this, new EventArgs());
                    }
                }
                else
                {
                    this.dirty = false;
                    
                    foreach (SavableProperties dependant in this.dependents)
                    {
                        dependant.IsDirty = false;
                    }
                    
                    if (this.Clean != null) 
                    {
                        this.Clean(this, new EventArgs());
                    }
                }
            }
        }

        protected void AddDependent(SavableProperties subject)
        {
            subject.Dirty += this.OnDependentDirty;
            this.dependents.Add(subject);
        }

        protected void RemoveDependent(SavableProperties subject)
        {
            subject.Dirty -= this.OnDependentDirty;
            this.dependents.Remove(subject);
        }

        protected virtual void OnPropertyChanged()
        {
            if (this.PropertyChanged != null) 
            {
                this.PropertyChanged(this, new PropertyChangedEventArgs("sp"));
            }
            
            this.IsDirty = true;
        }
        
        private void OnDependentDirty(object source, EventArgs args)
        {
            this.IsDirty = true;
        }
    }
}